### ISSUE TITLE
MongoDB single date/month field filters aren't working correctly

### ISSUE THREAD

**Issue by dacort** [metabase/metabase#11715]
**Describe the bug**
MongoDB single date/month field filters aren't working correctly. This is similar to #11690 just expanded to year/month.

**Logs**
When filtering a date or timestamp by "this month" or "this year" or "last month/year", the mongo query generated does not include the last day of the month/year. 

Example generated query for "last month":
```json
"$match" : {
    "$and" : [
        {
            "birthdate" : {
                "$gte" : ISODate("2019-12-01T08:00:00Z")
            }
        },
        {
            "birthdate" : {
                "$lt" : ISODate("2019-12-31T08:00:00Z")
            }
        }
    ]
}
```

Example generated query for "this month":
```json
"$match" : {
    "$and" : [
        {
            "birthdate" : {
                "$gte" : ISODate("2020-01-01T08:00:00Z")
            }
        },
        {
            "birthdate" : {
                "$lt" : ISODate("2020-01-31T08:00:00Z")
            }
        }
    ]
}
```

**Expected behavior**
Full month/year is included

**Screenshots**
This came up when querying a date field using a native query with a field filter. 

Native query: `[ {$match: {{date_var}} } ]`

**Information about your Metabase Installation:**

```json
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.117 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.2+9",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "11.0.2",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.2+9",
    "os.name": "Mac OS X",
    "os.version": "10.14.6",
    "user.language": "en",
    "user.timezone": "America/Los_Angeles"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "mongo"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "run-mode": "prod",
    "version": {
      "date": "2020-01-13",
      "tag": "v0.34.1",
      "branch": "release-0.34.x",
      "hash": "265695c"
    },
    "settings": {
      "report-timezone": "US/Pacific"
    }
  }
}
```

**Severity**
Fairly high - results in inaccurate results.

**Additional context**
n/a

**Comment by dacort** [comment]
A couple tests for what I think would be the correct behavior

```clojure
(testing "relative year range"
    (mt/with-clock #t "2019-12-13T12:00:00.000Z[UTC]"
                   (is (= "[{$match: {$and: [{\"date\": {$gte: ISODate(\"2018-01-01\")}}, {\"date\": {$lt: ISODate(\"2019-01-01\")}}]}}]"
                          (substitute {:date (field-filter "date" :date/range "lastyear")}
                                      ["[{$match: " (param :date) "}]"])))))
  (testing "relative month range"
    (mt/with-clock #t "2019-12-13T12:00:00.000Z[UTC]"
                   (is (= "[{$match: {$and: [{\"date\": {$gte: ISODate(\"2018-11-01\")}}, {\"date\": {$lt: ISODate(\"2018-12-01\")}}]}}]"
                          (substitute {:date (field-filter "date" :date/range "lastmonth")}
                                      ["[{$match: " (param :date) "}]"])))))
```

**Comment by camsaul** [comment]
Fixed by #11691
